### PR TITLE
Add HunyuanVideo-1.5 contrib model

### PR DESCRIPTION
**Pull request by jimburtoft**
*Note: The below template includes items meant for model contributions only. For other contributions such as bug fixes, features, etc., only fill out the relevant portions of the form.*

## Description

HunyuanVideo-1.5 text-to-video generation on AWS Trainium. 8.33B DiT transformer with TP=4 parallelism and NKI flash attention, producing 480x848 video at ~55s with Classifier-Free Guidance.

Multi-component pipeline: DiT backbone (Neuron TP=4), 3D Causal VAE (Neuron tiled decode), byT5 glyph encoder (Neuron traced), Qwen2.5-VL 7B LLM text encoder (CPU bf16).

## Model Information

**Model Name:** HunyuanVideo-1.5 (tencent/HunyuanVideo-1.5)

**Model Architecture:** 54 double-stream DiT blocks, hidden_size=2048, 16 attention heads, head_dim=128, flow matching with Euler scheduler

**Purpose:** Text-to-video generation (480x848 resolution, 5 frames)

### Required Components

- [x] **Accuracy Test** (`test/integration/test_model.py`)
  - byT5 encoder + mapper: cosine similarity >= 0.999 vs CPU reference (measured: 1.000 / 0.9999)
  - VAE tile decode: finite output, correct shape (16x spatial upsampling)
  - E2E pipeline: produces 5 frames at 480x848 resolution
  - E2E performance: 50-step generation within 120s wall-clock

- [x] **README.md** with the following sections:
  - [x] **Usage Example**: Full setup from scratch with environment variables
  - [x] **Compatibility Matrix**: Validated on trn2.3xlarge with SDK 2.28
  - [x] **Example Checkpoints**: Links to all 3 required HuggingFace models
  - [x] **Testing Instructions**: Commands for component and E2E test groups

- [x] **Source Code** (`src/`)
  - 8 Python modules covering DiT TP wrapper, CPU preprocessor, VAE compilation, tiled VAE decode, byT5 tracing, negative embedding caching, E2E pipeline, and DiT compilation
  - Uses `torch_neuronx.trace()` + `parallel_model_trace()` for TP=4

### Optional Components

- [ ] **Unit Tests** (CPU or Neuron-based)

## Folder Structure

Confirm your contribution follows this structure:

```
/contrib/models/HunyuanVideo-1.5/
  README.md
  /src
    __init__.py
    dit_tp_wrapper.py
    dit_wrapper.py
    e2e_pipeline.py
    compile_vae_neuron.py
    tiled_vae_decode.py
    trace_byt5.py
    cache_neg_embeddings.py
    recompile_dit_masked.py
  /test
    __init__.py
    /integration
      __init__.py
      test_model.py
  /samples
    /neuron
      frame_0000.png
      frame_0002.png
      frame_0004.png
  /examples
    generate_video.py
```

## Testing

**How did you test this change?**

All tests executed on trn2.3xlarge (LNC=2, 4 logical NeuronCores) with pre-compiled models.

Component tests (byT5 + VAE) run in-process; E2E tests run as subprocess.
Tests must be run in two groups to avoid NeuronCore contention:

```bash
pytest test/integration/test_model.py -k "ByT5 or VAE" -v  # 2 passed, 54s
pytest test/integration/test_model.py -k "E2E" -v           # 2 passed, 150s
```

**Test Results:**

| Test | Result | Key Metric |
|------|--------|------------|
| byT5 encoder accuracy | PASSED | cos_sim = 1.000001 |
| byT5 mapper accuracy | PASSED | cos_sim = 0.999947 |
| VAE tile decode | PASSED | [1,3,5,128,128], 177ms, all finite |
| E2E frame generation (2 steps) | PASSED | 5 frames at 480x848, 23.2s |
| E2E performance (50 steps) | PASSED | 327ms/step, 60.8s wall-clock |

## Compatibility

**Tested with:**
- **Neuron SDK Version(s):** 2.28 (neuronx-cc 2.22, torch-neuronx 2.9.0)
- **Instance Type(s):** trn2.3xlarge (LNC=2)
- **PyTorch Version:** 2.9.0
- **Python Version:** 3.12.3

## Additional Information

**5 code patches required for Neuron compatibility:**
1. Dtype mismatch in token_refiner (mask.float() -> mask.to(x.dtype))
2. FlexAttention replaced with F.scaled_dot_product_attention
3. NaN from SDPA mask expansion (key-only mask fix)
4. byT5 output dtype cast to bf16
5. RoPE rotate_half: interleaved-pair rotation (critical for visual quality)

**Performance:**
- DiT per-step: 327ms (TP=4, NKI flash attention)
- VAE tiled decode: 8.5s (45 tiles x 177ms)
- byT5 encode: 4.4ms
- E2E with CFG (50 steps): ~55s
- E2E without CFG: ~25s

**Known limitations:**
- LLM (Qwen2.5-VL 7B) runs on CPU (~14s) -- too large for single-core Neuron trace
- VAE requires monkey-patches for replication_pad3d compiler issue
- B=2 batched CFG is 4.2x slower than sequential (HBM bandwidth saturation)

## Related Issues

N/A

## vLLM Integration

- [ ] This model/feature is intended for use with vLLM
- [ ] Documentation includes vLLM registration instructions

---

**By submitting this PR, I confirm that:**
- [x] I have read and followed the [contributing guidelines](../contrib/CONTRIBUTING.md)
- [x] This is a community contribution and may have limited testing compared to officially-supported models
- [x] The code follows best practices and is well-documented
- [x] All required components listed above are included

### CODE REVIEW

File: contrib/models/HunyuanVideo-1.5/src/cache_neg_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    python ./cache_neg_embeddings.py
+"""
+
+import os
+import sys
+import time
+import torch
+
+sys.path.insert(0, os.environ.get("HUNYUAN_REPO_DIR", "./HunyuanVideo-1.5"))
+os.environ["HUNYUAN_ATTN_MODE"] = "torch"
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/cache_neg_embeddings.py:23]: Use os.environ.setdefault() instead of setting directly or set environment variables from outside scripts -- this might cause different scripts to use different values.

File: contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+import os
+import time
+import json
+import argparse
+
+# Set env BEFORE imports
+os.environ["NEURON_RT_NUM_CORES"] = "4"
+os.environ["XLA_DISABLE_FUNCTIONALIZATION"] = "1"
+os.environ["NEURON_RT_VIRTUAL_CORE_SIZE"] = "2"
+os.environ["NEURON_FUSE_SOFTMAX"] = "1"
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py:38]: Same comment as in cache_neg_embeddings.py

Use os.environ.setdefault() instead of setting directly or set environment variables from outside scripts -- this might cause different scripts to use different values.

File: contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+            mp4_path = f"{output_dir}/output.mp4"
+            writer = imageio.get_writer(
+                mp4_path, fps=8, codec="libx264", quality=8, pixelformat="yuv420p"
+            )
+            for fi in range(T_out_actual):
+                frame_np = frames_uint8[:, fi].permute(1, 2, 0).cpu().numpy()
+                writer.append_data(frame_np)
+            writer.close()
+            print(f"  Saved {mp4_path} ({T_out_actual} frames, 8 fps)")
+        except Exception as e:
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py:872]: It would be better to use more specific exception handling.

File: contrib/models/HunyuanVideo-1.5/src/compile_vae_neuron.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+                    compiler_args=compiler_args,
+                    compiler_workdir=args.compiler_workdir + "_4x4",
+                )
+                compile_time = time.time() - t2
+                print(f"4x4 tile traced in {compile_time:.1f}s")
+                tile_h = tile_h_small
+                tile_w = tile_w_small
+                z_test = z_small
+                cpu_out = cpu_out_small
+            except Exception as e2:
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/compile_vae_neuron.py:214]: Same comment about being specific about Exceptions, especially if they are nested Exceptions

File: contrib/models/HunyuanVideo-1.5/src/recompile_dit_masked.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            checkpoint=remapped_sd, model=model, serialize_path=weights_path
+        )
+
+    # Load and benchmark
+    print("\nLoading to Neuron...")
+    t0 = time.time()
+    sharded_weights = []
+    for rank in range(TP_DEGREE):
+        sharded_weights.append(
+            load_file(f"{weights_path}/tp{rank}_sharded_checkpoint.safetensors")
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/recompile_dit_masked.py:111]: It would be better to check for the existence of these files before loading

File: contrib/models/HunyuanVideo-1.5/src/recompile_dit_masked.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    sep = "=" * 60
+    print(sep)
+    print("Recompile DiT with text_mask KV zeroing fix")
+    print(sep)
+
+    # Load checkpoint
+    ckpt_path = os.path.join(MODEL_DIR, "diffusion_pytorch_model.safetensors")
+    print(f"Loading checkpoint...")
+    t0 = time.time()
+    original_sd = load_file(ckpt_path)
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/src/recompile_dit_masked.py:41]: Same comment about loading file without checking for existence first

File: contrib/models/HunyuanVideo-1.5/README.md [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+## Usage
+
+### 1. Prerequisites
+
+```bash
+# SSH into trn2.3xlarge instance
+source /opt/aws_neuronx_venv_pytorch_inference_vllm_0_13/bin/activate
+
+# Clone HunyuanVideo repo (needed for model definitions)
+git clone https://github.com/Tencent/HunyuanVideo-1.5.git
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/README.md:50]: Wrong Github URL

File: contrib/models/HunyuanVideo-1.5/README.md [hunk 8]
```diff
@@ -0,0 +1,334 @@
+# Contrib Model: HunyuanVideo-1.5
```
Comment (lutfanm-aws) [inline contrib/models/HunyuanVideo-1.5/README.md:1]: A dependencies section, or a requirements.txt file would help agents install necessary dependencies before attempting compilation and inference.

**Review by lutfanm-aws — APPROVED** [review]
Comments posted are non-blocking because I managed to compile and generate an image with the modeling files here.

File: contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py (hunk 3 above)
Comment (tejasamx-aws) [reply contrib/models/HunyuanVideo-1.5/src/e2e_pipeline.py:872]: +1

**Review by tejasamx-aws — APPROVED** [review]
